### PR TITLE
ENH: `__array_function__` support for `np.lib`, part 2/2

### PR DESCRIPTION
**Pull request by shoyer**
xref #12028

np.lib.npyio through np.lib.ufunclike

### CODE REVIEW

**Comment by shoyer** [comment]
There's some weird interaction between `iscomplexobj` and `assert_equal` (which internally calls `iscomplexobj`) that is causing the observed test failures, e.g.,
```

self = <numpy.core.tests.test_overrides.TestArrayFunctionDispatch object at 0x7f8910047ba8>
    def test_interface(self):
    
        class MyArray(object):
            def __array_function__(self, func, types, args, kwargs):
                return (self, func, types, args, kwargs)
    
        original = MyArray()
        (obj, func, types, args, kwargs) = dispatched_one_arg(original)
        assert_(obj is original)
        assert_(func is dispatched_one_arg)
        assert_equal(set(types), {MyArray})
>       assert_equal(args, (original,))
MyArray    = <class 'numpy.core.tests.test_overrides.TestArrayFunctionDispatch.test_interface.<locals>.MyArray'>
args       = (<numpy.core.tests.test_overrides.TestArrayFunctionDispatch.test_interface.<locals>.MyArray object at 0x7f8910047898>,)
func       = <function dispatched_one_arg at 0x7f8913644e18>
kwargs     = {}
obj        = <numpy.core.tests.test_overrides.TestArrayFunctionDispatch.test_interface.<locals>.MyArray object at 0x7f8910047898>
original   = <numpy.core.tests.test_overrides.TestArrayFunctionDispatch.test_interface.<locals>.MyArray object at 0x7f8910047898>
self       = <numpy.core.tests.test_overrides.TestArrayFunctionDispatch object at 0x7f8910047ba8>
types      = [<class 'numpy.core.tests.test_overrides.TestArrayFunctionDispatch.test_interface.<locals>.MyArray'>]
../builds/venv/lib/python3.6/site-packages/numpy/core/tests/test_overrides.py:190: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../builds/venv/lib/python3.6/site-packages/numpy/core/overrides.py:151: in public_api
    implementation, public_api, relevant_args, args, kwargs)
../builds/venv/lib/python3.6/site-packages/numpy/core/overrides.py:96: in array_function_implementation_or_override
    return implementation(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
x = 5
    @array_function_dispatch(_iscomplexobj_dispatcher)
    def iscomplexobj(x):
        """
        Check for a complex type or an array of complex numbers.
    
        The type of the input is checked, not the value. Even if the input
        has an imaginary part equal to zero, `iscomplexobj` evaluates to True.
    
        Parameters
        ----------
        x : any
            The input can be of any type and shape.
    
        Returns
        -------
        iscomplexobj : bool
            The return value, True if `x` is of a complex type or has at least
            one complex element.
    
        See Also
        --------
        isrealobj, iscomplex
    
        Examples
        --------
        >>> np.iscomplexobj(1)
        False
        >>> np.iscomplexobj(1+0j)
        True
        >>> np.iscomplexobj([3, 1+0j, True])
        True
    
        """
        try:
>           dtype = x.dtype
E           RecursionError: maximum recursion depth exceeded
x          = 5
../builds/venv/lib/python3.6/site-packages/numpy/lib/type_check.py:311: RecursionError
```

**Comment by shoyer** [comment]
The problem seems to be that my dummy `__array_function__` implementation returns `self` for all functions, which then breaks `assert_equal` when `iscomplexobj()` is overloaded.

I see three possible ways to fix this:
- adjust my `__array_function__` implementation to special case `iscomplexobj()`
- use `assert_(args == (original,))` instead of `assert_equal()`
- add more fallback logic to `assert_equal()` to handle cases where `iscomplexobj()` returns a nonsensical result (i.e., a non-boolean)

**Comment by shoyer** [comment]
I decided to go for a mix of 2 and 3:
- `assert_equal` should be OK with `np.iscomplexobj` raising `TypeError`.
- for the one case where I define `__array_function__` for everything, I use `assert_` instead of `assert_equal`.

File: numpy/lib/ufunclike.py
Comment (hameerabbasi) [inline numpy/lib/ufunclike.py]: Should this be added to the dispatcher?

**Review by hameerabbasi — APPROVED** [review]
We perhaps should get rid of the hack sometime, but other than that I'm okay with this.

File: numpy/lib/ufunclike.py
Comment (shoyer) [reply numpy/lib/ufunclike.py]: Yes, this is a better idea. That way the warning will be raised even if the function is dispatching.

**Comment by shoyer** [comment]
Note: This one still needs more work before it's ready to merge (to fix the deprecation warnings).

**Comment by shoyer** [comment]
OK, inspired by @mhvk's and @hameerabbasi's comments I consolidated a bunch of unnecessarily repeated dispatcher functions here, too.

**Comment by shoyer** [comment]
This is ready for a final review.

**Review by hameerabbasi — CHANGES_REQUESTED** [review]
One minor bug I spotted as I was going through this.

File: numpy/lib/scimath.py
Comment (shoyer) [reply numpy/lib/scimath.py]: yes -- I think I missed this because I only read the docstring, which says `n` is an int.

File: numpy/lib/shape_base.py
Comment (mhvk) [inline numpy/lib/shape_base.py]: Here you could use a single `_split_dispatcher` with the function above as well as below (where the function currently gets redefined, but then it does get reused a few times).

File: numpy/lib/twodim_base.py
Comment (mhvk) [inline numpy/lib/twodim_base.py]: Maybe call it `triul_dispatcher`, since it is used for both?

File: numpy/lib/twodim_base.py
Comment (mhvk) [inline numpy/lib/twodim_base.py]: Same, either `_tri` or `_triul`

File: numpy/lib/type_check.py
Comment (mhvk) [inline numpy/lib/type_check.py]: Next few could be common `_is_type_dispatcher`?

File: numpy/testing/_private/utils.py
Comment (mhvk) [inline numpy/testing/_private/utils.py]: Given the comment above, I think this should not be changed... Alternatively, at least the comment should be adjusted.

**Review by mhvk — COMMENTED** [review]
Looks good but some suggestions for using decorators for a group.

Also, more important, the last comment about the change in testing `utils.py` - that was really tricky to get right; I would suggest not to change it unless truly needed.

File: numpy/lib/shape_base.py
Comment (shoyer) [reply numpy/lib/shape_base.py]: good catch -- fixed

File: numpy/lib/twodim_base.py
Comment (shoyer) [reply numpy/lib/twodim_base.py]: done

File: numpy/lib/twodim_base.py
Comment (shoyer) [reply numpy/lib/twodim_base.py]: done

File: numpy/lib/type_check.py
Comment (shoyer) [reply numpy/lib/type_check.py]: done!

File: numpy/testing/_private/utils.py
Comment (shoyer) [reply numpy/testing/_private/utils.py]: Good catch. The problem is that now `np.all()` may likely not be implemented for an ndarray subclass, if it doesn't define `np.all` in `__array_function__`.

I think we can just delete this comment, because it's no longer true that `np.all()` can work when a subclass implements`.all()` differently -- we already use `getattr()` to pull out a `all()` method to call if one exists.

Comment (mhvk) [reply numpy/testing/_private/utils.py]: The reason I worry is that this change is a very recent one, by @charris (https://github.com/numpy/numpy/pull/11756), which correctly something that `.all()` didn't catch. (Sorry, have to run, not sure what the real issue was...)

Comment (shoyer) [reply numpy/testing/_private/utils.py]: OK, I see, thanks. I think the comment was a little misleading here -- I think the problem was classes that define equality differently (e.g., to return a boolean) rather than classes that don't define an `.all()` method.

File: numpy/testing/_private/utils.py
Comment (hameerabbasi) [inline numpy/testing/_private/utils.py]: Will this pass through subclasses? And if so, will that be an issue?

Comment (mhvk) [reply numpy/testing/_private/utils.py]: Agree with the worry of @hameerabbasi - also, at least the comment reads a bit weird now: I don't think we have to worry about classed that define `__array_function__` but do not override `np.all` - they're new and they can override easily as part of their trials; maybe just leave it as it was?

Comment (shoyer) [reply numpy/testing/_private/utils.py]: Yes, this passes through subclasses -- which is the point.

> I don't think we have to worry about classed that define __array_function__ but do not override np.all - they're new and they can override easily as part of their trials; maybe just leave it as it was?

I don't exactly disagree with this, but this did come up with the ndarray subclass I made for unit testing purposes. With the arrival of `__array_function__`, it will be easier to depend on built-in methods like `all()` rather than `np.all()`. I guess we offer no guarantees for subclasses that don't implement the full numpy API, but it still feels like not a great user experience.

Two other ways to do this:
1. Cast to a NumPy boolean scalar/array, which guarantees that we have an `all()` method: `np.bool_(x_id == y_id).all()`. This looks a little less hacky.
2. Give up on duck-typing for `assert_array_equal`, and instead decorate it with `array_function_dispatch` for explicit dispatching.

Comment (hameerabbasi) [reply numpy/testing/_private/utils.py]: In that case, if this won't cause any problems, then I agree with this design.

Comment (mhvk) [reply numpy/testing/_private/utils.py]: Option 1 seems to cover all bases, and is shorter than what you have now.

Comment (shoyer) [reply numpy/testing/_private/utils.py]: Hmm. Testing out masked arrays, they no longer seem to return masked arrays from `.all()` or `np.all()`:
```
In [16]: x = np.ma.array([1, 2, 3], mask=[False, True, False])

In [17]: x == x
Out[17]:
masked_array(data=[True, --, True],
             mask=[False,  True, False],
       fill_value=True)

In [18]: (x == x).all()
Out[18]: True
```

Comment (mattip) [reply numpy/testing/_private/utils.py]: That should cause a test failure, and if there is no test like that we should add one. Also note there are test failures with datetime64

Comment (hameerabbasi) [reply numpy/testing/_private/utils.py]: > That should cause a test failure, and if there is no test like that we should add one. Also note there are test failures with datetime64

I'm not sure. What would you do with a mask at all on a reduction? I mean, it makes sense to only reduce over the non-masked objects, but what is the output mask? For zero nonmasked elements do you set it to the identity of the ufunc or to `False`?

**Comment by mhvk** [comment]
Wow, that should cover it! Thanks, @shoyer!

**Comment by shoyer** [comment]
OK, I plan to merge this shortly unless I get more feedback.

**Pull request merged**
